src/core/connection.c

```diff
@@ -1137,56 +1137,9 @@ QuicConnTimerSetEx(
         (uint8_t)Type,
         Delay);
 
-    //
-    // Find the current and new index in the timer array for this timer.
-    //
-
-    uint32_t NewIndex = ARRAYSIZE(Connection->Timers);
-    uint32_t CurIndex = 0;
-    for (uint32_t i = 0; i < ARRAYSIZE(Connection->Timers); ++i) {
-        if (Connection->Timers[i].Type == Type) {
-            CurIndex = i;
-        }
-        if (i < NewIndex &&
-            NewExpirationTime < Connection->Timers[i].ExpirationTime) {
-            NewIndex = i;
-        }
-    }
-
-    if (NewIndex < CurIndex) {
-        //
-        // Need to move the timer forward in the array.
-        //
-        CxPlatMoveMemory(
-            Connection->Timers + NewIndex + 1,
-            Connection->Timers + NewIndex,
-            sizeof(QUIC_CONN_TIMER_ENTRY) * (CurIndex - NewIndex));
-        Connection->Timers[NewIndex].Type = Type;
-        Connection->Timers[NewIndex].ExpirationTime = NewExpirationTime;
-
-    } else if (NewIndex > CurIndex + 1) {
-        //
-        // Need to move the timer back in the array. Ignore changes that
-        // wouldn't actually move it at all.
-        //
-        CxPlatMoveMemory(
-            Connection->Timers + CurIndex,
-            Connection->Timers + CurIndex + 1,
-            sizeof(QUIC_CONN_TIMER_ENTRY) * (NewIndex - CurIndex - 1));
-        Connection->Timers[NewIndex - 1].Type = Type;
-        Connection->Timers[NewIndex - 1].ExpirationTime = NewExpirationTime;
-    } else {
-        //
-        // Didn't move, so just update the expiration time.
-        //
-        Connection->Timers[CurIndex].ExpirationTime = NewExpirationTime;
-        NewIndex = CurIndex;
-    }
-
-    if (NewIndex == 0 || CurIndex == 0) {
-        //
-        // The first timer was updated, so make sure the timer wheel is updated.
-        //
+    Connection->ExpirationTimes[Type] = NewExpirationTime;
+    if (Connection->EarliestExpirationTime > NewExpirationTime) {
```

Comment:
What if you changed the previous earliest time? You need to update that here too then.

Reply:
You meant when the timer type being set happens to hold the earliest expiration time and we are setting it to a later time (setting it to an earlier time is covered by the if check)?

That case is handled in a lazy manner. Since the earliest exp time is not updated, it's possible that we end up calling QuicConnTimerExpired for that stale earliest exp time. However, since the value's guaranteed to be earlier than any scheduled timers on this connection, there will be no side-effect, and QuicConnTimerExpired will then update the timer wheel with up-to-date earliest exp time. I guess we don't need to have tricky stuff like this.